### PR TITLE
[NTOS:PNP][SDK:LIB] Build assignment ordering for arbiters

### PR DESCRIPTION
**Pull request by vgalnt**
## Purpose

Implement ArbBuildassignmentOrdering() and dependent functions for the arbiter library and PnP arbiters.

## Proposed changes

- [LIB:ARBITER] Implemented:
   ArbInitializeOrderingList(),
   ArbFreeOrderingList(),
   ArbAddOrdering(),
   ArbPruneOrdering(),
   ArbBuildAssignmentOrdering().

- [LIB:ARBITER] Added ArbpGetRegistryValue().

- [NTOS:PNP] Implemented:
   IopIrqUnpackRequirement(),
   IopDmaUnpackRequirement(),
   IopGenericUnpackRequirement(),
   IopIrqTranslateOrdering(),
   IopGenericTranslateOrdering().

### CODE REVIEW

**Comment by binarymaster** [comment]
Cc @disean - might want to review this too

**Comment by vgalnt** [comment]
File with logs for different HALs - [CORE-17996](https://jira.reactos.org/browse/CORE-17996?focusedCommentId=132312&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-132312)

**Comment by Extravert-ir** [comment]
Can you explain please, what is "ordering" in the context of an arbiter?
Also, reviewing things with a brief explanation of what each of the Arb function does would be a lot more easy. I'm fine if it would be in Russian even.

**Comment by Extravert-ir** [comment]
By the way, what's the point in *UnpackRequirement functions? They seem to just extract 2 values from a resource descriptor. What are they supposed to do?

File: ntoskrnl/io/pnpmgr/arbiters.c [hunk 1]
```diff
@@ -232,10 +282,23 @@ IopDmaUnpackRequirement(
...
+    DPRINT("DmaUnpackIo: [%p] Min %X Max %X\n",
+            IoDescriptor,
+            IoDescriptor->u.Dma.MinimumChannel,
+            IoDescriptor->u.Dma.MaximumChannel);
+
+    *OutMinimumChannel = IoDescriptor->u.Dma.MinimumChannel;
+    *OutMaximumChannel = IoDescriptor->u.Dma.MaximumChannel;
+
+    *OutParam1 = 1;
+    *OutParam2 = 1;
```
Comment (Extravert-ir) [inline ntoskrnl/io/pnpmgr/arbiters.c:299]: It seems like we know now that these are "Length" and "Alignment". But why assigning "1" to them, what is that supposed to mean?

File: ntoskrnl/io/pnpmgr/arbiters.c [hunk 2]
```diff
@@ -331,8 +394,34 @@ IopGenericUnpackRequirement(
...
+           IoDescriptor->Type == CmResourceTypeMemory);
+
+    *OutLength = IoDescriptor->u.Generic.Length;
+    *OutAlignment = IoDescriptor->u.Generic.Alignment;
+
+    *OutMinimumAddress = IoDescriptor->u.Generic.MinimumAddress.QuadPart;
+    *OutMaximumAddress = IoDescriptor->u.Generic.MaximumAddress.QuadPart;
+
+    if (IoDescriptor->u.Generic.Alignment == 0)
+        *OutAlignment = 1;
```
Comment (Extravert-ir) [inline ntoskrnl/io/pnpmgr/arbiters.c:414]: Why is this needed?

File: ntoskrnl/io/pnpmgr/arbiters.c [hunk 3]
```diff
@@ -331,8 +394,34 @@ IopGenericUnpackRequirement(
...
+
+    if (IoDescriptor->Type == CmResourceTypeMemory &&
```
Comment (Extravert-ir) [inline ntoskrnl/io/pnpmgr/arbiters.c:416]: Add a comment here, something like
```
// Fixup memory requirements for devices which support only 24bit addresses
```

**Comment by vgalnt** [comment]
You can study how the arbiter works using WinDbg ("!arbiter" command)
[Example from CORE-17996](https://jira.reactos.org/browse/CORE-17996?focusedCommentId=132371&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-132371)
Also useful [hardware resources](https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/hardware-resources#)

I rarely use comments (very often they just get in the way, but sometimes they are really very useful), I try to understand how the code works from the source using DPRINT(1)s. Almost every function has at least one DPRINT. You can just do "//#define NDEBUG" in arbiter.c and arbiters.c and get a verbose log file.

**Comment by vgalnt** [comment]
struct _ARBITER_INSTANCE имеет OrderingList и ReservedList, которые заполняются из реестра (CurrentControlSet\Control\Arbiters\AllocationOrder). Они упорядочены по приоритетности альтернатив, т.е. приоритет каждой строки-альтернативы уменьшается. Они участвуют только в функции ArbGetNextAllocationRange(), до которой ещё, видимо, не скоро, потому что основной функционал в диспетчере ресурсов PnP менеджера, на него нужны месяцы ...

**Comment by vgalnt** [comment]
The *Unpack* and *Pack* (as well as the rest) functions that are part of ARBITER_INSTANCE are defined in Windows... This interface should be compatible whether we like it or not :)

File: ntoskrnl/io/pnpmgr/arbiters.c (hunk 1 above)
Comment (vgalnt) [reply ntoskrnl/io/pnpmgr/arbiters.c:299]: The "Length" and "Alignment" values will be set to "1". 

Вот, например, код из отсутствующей функции (которая ещё в разработке)
```c
    Status = Arbiter->UnpackRequirement(IoDescriptor,
                                        &Alternative->Minimum,
                                        &Alternative->Maximum,
                                        &Alternative->Length,
                                        &Alternative->Alignment);
```

File: ntoskrnl/io/pnpmgr/arbiters.c (hunk 2 above)
Comment (vgalnt) [reply ntoskrnl/io/pnpmgr/arbiters.c:414]: Default alignment will be set to one byte.

File: ntoskrnl/io/pnpmgr/arbiters.c (hunk 3 above)
Comment (vgalnt) [reply ntoskrnl/io/pnpmgr/arbiters.c:417]: Ok
